gammapy/stats/poisson.py

```diff
@@ -413,6 +414,84 @@ def _significance_direct_on_off(n_on, n_off, alpha):
 
     return significance
 
+def Helene_ULs(excess, error, conf_level=95.45):
+    """Compute Upper limit based on Helene et al., NIM 212 (1983) 319
+
+    Parameters
+    ----------
+    excess : double
+        Signal excess
+    error : double
+        Gaussian Signal error
+    conf_level: double
+        Confidence level in percentage (1sigma=68.27, 2sigma=95.45, 3sigma=99.73, 5sigma=99.999943)
+
+    Returns
+    -------
+    double
+        Upper limit for the excess
+    """
+
+    conf_level1 = 1. - conf_level/100.
+
+    if error <= 0:
+        return 0.
+
+    from math import sqrt
+    from scipy.special import erf
+    if excess >= 0.:
+        zeta = excess/error
+        value = zeta/sqrt(2.)
+        integral = (1.+erf(value))/2.
+        integral2 = 1. - conf_level1*integral
+        value_old = value
+
+        # The 1st Loop is for Speed & 2nd For Precision
+        value_new = value_old+0.01
+        if integral > integral2:
+            value_new = 0.
+        integral = (1.+erf(value_new))/2.
+
+        while integral < integral2:
+            value_old = value_new
+            value_new = value_new+0.01
+            integral = (1.+erf(value_new))/2.
+        value_new = value_old+0.0000001
+        integral = (1. + erf(value_new))/2.
+
+        while integral < integral2:
+            value_old = value_new
```

Comment:
PyCharm shows me that this `value_old` is unused; you can delete this line and the result will be the same.

This code is really hard to understand.
It looks like the excess > 0 and < 0 are the same or a similar algorithm?
If yes, can you put the algorithm in a helper functino and call it twice, instead of duplicating the ~ 10 lines of the algorithm?

Reply:
The comment on value_old is accepted and its removal is done.
For the different cases on excess, the algo is slightly different and a difference should be made. I have however simplify a bit the code as there are common ligns. Concerning its complexity, the code is coming from a C lib and optimised in this respect. I have not changed it and I do not plan to optimise it in a short term.